theano/scan_module/scan_utils.py

```diff
@@ -55,9 +55,24 @@ def safe_new(x, tag='', dtype=None):
             return nwx
         else:
             return x.clone()
-
-    # at this point we should only have Variables
-    assert isinstance(x, theano.Variable)
+    # Note, as_tensor_variable will convert the Scalar into a
+    # TensorScalar that will require a ScalarFromTensor op,
+    # making the pushout optimization fail
+    elif isinstance(x, scalar.ScalarVariable):
+        if dtype:
+            nw_x = scalar.get_scalar_type(dtype=dtype)()
+        else:
+            nw_x = x.type()
+        nw_x.name = nw_name
+        return nw_x
+    else:
+        try:
+            x = tensor.as_tensor_variable(x)
+        except TypeError:
+            # This could happen for example for random states, and I really
+            # want to avoid the convoluted logic that checks for cuda
+            # ndarrays
+            pass
```

Comment:
We are trying to understand why this cause this. We have hypothesis, but we must check them before. This part was there to fix some strange segfault, so just a revert isn't the ideal fix.